File: 2022/go/05.go
package main

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"unicode"
)

type Day5 struct {
	stacks    []*stack
	nStacks   []*nStack
	movements []*movement
}

func (*Day5) Day() int {
	return 5
}

func (puzzle *Day5) Run(input string) {
	puzzle.parseInput(input)
	puzzle.part1()
	puzzle.part2()
}

func (puzzle *Day5) part1() {
	for _, movement := range puzzle.movements {
		for i := 0; i < movement.quantity; i++ {
			c := puzzle.stacks[movement.fromIndex].pop()
			puzzle.stacks[movement.toIndex].push(c)
		}
	}

	fmt.Print("Part1: ")

	for _, stack := range puzzle.stacks {
		fmt.Printf("%c", stack.peek())
	}

	fmt.Println()
}

func (puzzle *Day5) part2() {
	for _, movement := range puzzle.movements {
		c := puzzle.nStacks[movement.fromIndex].pop(movement.quantity)
		puzzle.nStacks[movement.toIndex].push(c)
	}

	fmt.Print("Part2: ")

	for _, stack := range puzzle.nStacks {
		fmt.Printf("%c", stack.peek(1)[0])
	}

	fmt.Println()
}

func (puzzle *Day5) parseInput(input string) {
	isParsingStacks := true

	for _, line := range strings.Split(input, "\n") {
		line = strings.TrimRightFunc(line, unicode.IsSpace)

		if line == "" {
			isParsingStacks = false
			continue
		}

		if isParsingStacks {
			puzzle.parseCrateLine(line)
		} else {
			puzzle.parseMovement(line)
		}
	}

	for _, stack := range puzzle.stacks {
		stack.isReversed = false
		puzzle.nStacks = append(puzzle.nStacks, puzzle.cloneStackToNStack(stack))
	}
}

func (puzzle *Day5) parseCrateLine(line string) {
	if strings.HasPrefix(line, " 1 ") {
		return
	}

	stackIndex := 0

	runes := ([]rune)(line)

	for len(runes) > 0 {
		if len(puzzle.stacks) <= stackIndex {
			puzzle.stacks = append(puzzle.stacks, &stack{isReversed: true})
		}

		if runes[0] == '[' && runes[2] == ']' {
			c := runes[1]
			puzzle.stacks[stackIndex].push(c)
		}

		runes = runes[Min(4, len(runes)):]
		stackIndex++
	}
}

type movement struct {
	quantity  int
	fromIndex int
	toIndex   int
}

func (puzzle *Day5) parseMovement(line string) {
	regex, _ := regexp.Compile(`move (\d+) from (\d+) to (\d+)`)
	matches := regex.FindSubmatch(([]byte)(line))

	quantity, _ := strconv.Atoi((string)(matches[1]))
	from, _ := strconv.Atoi((string)(matches[2]))
	to, _ := strconv.Atoi((string)(matches[3]))

	puzzle.movements = append(puzzle.movements, &movement{
		quantity:  quantity,
		fromIndex: from - 1,
		toIndex:   to - 1,
	})
}

type stack struct {
	items      []rune
	isReversed bool
}

func (stack *stack) push(c rune) {
	if stack.isReversed {
		stack.items = append([]rune{c}, stack.items...)
	} else {
		stack.items = append(stack.items, c)
	}
}

func (stack *stack) peek() rune {
	if stack.isReversed {
		return stack.items[0]
	}
	return stack.items[len(stack.items)-1]
}

func (stack *stack) pop() rune {
	c := stack.peek()
	if stack.isReversed {
		stack.items = stack.items[1:]
	} else {
		stack.items = stack.items[:len(stack.items)-1]
	}
	return c
}

type nStack struct {
	items []rune
}

func (*Day5) cloneStackToNStack(stack *stack) *nStack {
	newArray := make([]rune, len(stack.items))

	for i := 0; i < len(newArray); i++ {
		newArray[i] = stack.items[i]
	}

	return &nStack{items: newArray}
}

func (stack *nStack) push(c []rune) {
	stack.items = append(stack.items, c...)
}

func (stack *nStack) peek(n int) []rune {
	return stack.items[len(stack.items)-n:]
}

func (stack *nStack) pop(n int) []rune {
	result := stack.peek(n)
	stack.items = stack.items[:len(stack.items)-n]
	return result
}

File: 2022/cs/05.cs
using System.Text.RegularExpressions;

namespace Day5Namespace;

record struct Movement(int SourceStackIndex, int TargetStackIndex, int Quantity);

class Day5 : IPuzzle
{
    public int Day { get; } = 5;

    private readonly List<Stack<char>> part1CrateStacks = new();
    private readonly List<Stack<char>> part2CrateStacks = new();
    private readonly List<Movement> movements = new();

    public void Run(string input)
    {
        ParseInput(input);

        Part1();
        Part2();
    }

    private void Part1()
    {
        foreach (Movement movement in movements)
        {
            for (int i = 0; i < movement.Quantity; i++)
            {
                char c = part1CrateStacks[movement.SourceStackIndex].Pop();
                part1CrateStacks[movement.TargetStackIndex].Push(c);
            }
        }

        Console.Write("Part1: ");

        foreach (Stack<char> crateStack in part1CrateStacks)
            Console.Write(crateStack.Peek());

        Console.WriteLine();
    }

    private void Part2()
    {
        Stack<char> tempStack = new();

        foreach (Movement movement in movements)
        {
            for (int i = 0; i < movement.Quantity; i++)
            {
                char c = part2CrateStacks[movement.SourceStackIndex].Pop();
                tempStack.Push(c);
            }

            while (tempStack.Count > 0)
                part2CrateStacks[movement.TargetStackIndex].Push(tempStack.Pop());
        }

        Console.Write("Part2: ");

        foreach (Stack<char> crateStack in part2CrateStacks)
            Console.Write(crateStack.Peek());

        Console.WriteLine();
    }

    private static void ParseCrateLine(string line, List<Stack<char>> crateStacks)
    {
        if (line.StartsWith(" 1 "))
            return;

        int stackIndex = 0;
        int lineIndex = 0;

        while (lineIndex < line.Length)
        {
            if (stackIndex >= crateStacks.Count)
                crateStacks.Add(new Stack<char>());

            if (line[lineIndex] == '[' && line[lineIndex + 2] == ']')
            {
                char c = line[lineIndex + 1];
                crateStacks[stackIndex].Push(c);
            }

            stackIndex++;
            lineIndex += 4;
        }
    }

    private static Regex MovementRegex = new Regex(@"move (\d+) from (\d+) to (\d+)", RegexOptions.Compiled);

    private void ParseMovementLine(string line)
    {
        Match m = MovementRegex.Match(line);

        int quantity = int.Parse(m.Groups[1].Value);
        int from = int.Parse(m.Groups[2].Value);
        int to = int.Parse(m.Groups[3].Value);

        movements.Add(new Movement(from - 1, to - 1, quantity));
    }

    private void ParseInput(string input)
    {
        List<Stack<char>> localCrateStacks = new();

        bool isParsingCrates = true;

        foreach (string line in input.Split('\n').Select(x => x.TrimEnd()))
        {
            if (line == string.Empty)
            {
                isParsingCrates = false;
                continue;
            }

            if (isParsingCrates)
                ParseCrateLine(line, localCrateStacks);
            else
                ParseMovementLine(line);
        }

        // Reconstruct stacks in proper order.
        foreach (Stack<char> localCrateStack in localCrateStacks)
        {
            Stack<char> crateStack = new();

            while (localCrateStack.Count > 0)
                crateStack.Push(localCrateStack.Pop());

            part1CrateStacks.Add(crateStack);
            part2CrateStacks.Add(new Stack<char>(crateStack.Reverse()));
        }
    }
}
